modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C

```diff
@@ -0,0 +1,635 @@
+      }
+    };
+    TIMPI::pull_parallel_vector_data(
+        _communicator, pull_requests, gather_functor, action_functor, &example);
+  }
+#else
+  mooseError("INSFVRhieChowInterpolator only supported for global AD indexing.");
+#endif
+}
+
+void
+INSFVRhieChowInterpolator::computeFirstAndSecondOverBars()
+{
+  _b2.reserve(_fe_problem.getEvaluableElementRange().size());
+
+  if (_standard_body_forces)
+  {
+    for (const auto & pr : _b)
+      _b2[pr.first] = pr.second;
+
+    return;
+  }
+
+  Moose::FV::interpolateReconstruct(_b2, _b, 1, false, _evaluable_fi, *this);
+}
+
+void
+INSFVRhieChowInterpolator::applyBData()
+{
+  const auto s = _sys.number();
+  for (auto * const elem : *_elem_range)
+  {
+    const auto elem_volume = _assembly.elementVolume(elem);
+    for (const auto i : index_range(_var_numbers))
+    {
+      const auto vn = _var_numbers[i];
+      // negative here because we swapped the sign in addToB and so now we need to swap it back
+      const auto residual = -elem_volume * _b2[elem->id()](i);
+      const auto dof_index = elem->dof_number(s, vn, 0);
+
+      if (_fe_problem.currentlyComputingJacobian())
+        _assembly.processDerivatives(residual, dof_index, _matrix_tags);
+      else
+        _assembly.processResidual(residual.value(), dof_index, _vector_tags);
+    }
+  }
+
+  if (_fe_problem.currentlyComputingJacobian())
+    _assembly.addCachedJacobian();
+  else
+    _assembly.addCachedResiduals();
+}
+
+void
+INSFVRhieChowInterpolator::finalizeBData()
+{
+#ifdef MOOSE_GLOBAL_AD_INDEXING
+  // First thing we do is to fill our _b map with 0's for any subdomains that do not have body force
+  // kernels active
+  std::set<SubdomainID> forceless_subs;
+  std::set_difference(_sub_ids.begin(),
+                      _sub_ids.end(),
+                      _sub_ids_with_body_forces.begin(),
+                      _sub_ids_with_body_forces.end(),
+                      std::inserter(forceless_subs, forceless_subs.end()));
+  for (const auto & elem : _mesh.active_local_subdomain_set_elements_ptr_range(forceless_subs))
+    _b[elem->id()] = 0;
+
+  // We do not have to push _b data because all that data should initially be
+  // local, e.g. we only loop over active local elements for FVElementalKernels
+
+  std::unordered_map<processor_id_type, std::vector<dof_id_type>> pull_requests;
+
+  for (auto * const elem : _fe_problem.getEvaluableElementRange())
+  {
+    const auto pid = elem->processor_id();
+    if (pid != this->processor_id())
+      pull_requests[pid].push_back(elem->id());
+  }
+
+  auto gather_functor = [this](const processor_id_type libmesh_dbg_var(pid),
+                               const std::vector<dof_id_type> & elem_ids,
+                               std::vector<VectorValue<ADReal>> & data_to_fill) {
+    mooseAssert(pid != this->processor_id(), "We shouldn't be gathering from ourselves.");
+    data_to_fill.resize(elem_ids.size());
+    for (const auto i : index_range(elem_ids))
```

Comment:
this only loops over 1 layer of neighbor elements right?

Reply:
1 would not be enough. It must be at least 2 for computation of third overbar of B. However, if you look at the logic right above the `gather_functor` you will see that we are pulling data for all evaluable elements that are not local elements. So if the user requested 3 layers of algebraic ghosting, then we would be doing pull requests for 3 ghost layers. Something I have labeled as a maybe to-do for this PR, and something that I think should actually be a mostly framework follow-on PR, is to make good utilities for creating element ranges that are like local + X layers which we don't have currently